### PR TITLE
Fix timestamp null value handling with grpc enabled.

### PR DESCRIPTION
**Pull request by qingling128**

### CODE REVIEW

File: lib/fluent/plugin/out_google_cloud.rb [hunk 1]
```diff
@@ -437,10 +437,15 @@ def write(chunk)
 
           ts_secs, ts_nanos = compute_timestamp(record, time)
           if @use_grpc
-            entry.metadata.timestamp = Google::Protobuf::Timestamp.new(
-              seconds: ts_secs,
-              nanos: ts_nanos
-            )
+            # If "seconds" or "nanos" are not integers (e.g. null), we will emit
```
Comment (andrewsg) [inline lib/fluent/plugin/out_google_cloud.rb]: I think "emit" here is intended to be "omit".

Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Oh god, right. Just changed this.

File: lib/fluent/plugin/out_google_cloud.rb [hunk 2]
```diff
@@ -437,10 +437,15 @@ def write(chunk)
 
           ts_secs, ts_nanos = compute_timestamp(record, time)
           if @use_grpc
-            entry.metadata.timestamp = Google::Protobuf::Timestamp.new(
-              seconds: ts_secs,
-              nanos: ts_nanos
-            )
+            # If "seconds" or "nanos" are not integers (e.g. null), we will omit
+            # the timestamp field and defer the decision of whether to set it to
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: s/of whether to set it to the current timestamp/on how too handle it/

File: lib/fluent/plugin/out_google_cloud.rb [hunk 3]
```diff
@@ -437,10 +437,15 @@ def write(chunk)
...
+            # the current timestamp to the downstream Logging API.
+            if ts_secs.is_a?(Integer) && ts_nanos.is_a?(Integer)
```
Comment (igorpeshansky) [inline lib/fluent/plugin/out_google_cloud.rb]: This seems to be a much broader check than what you're testing below. Shouldn't you also check non-integer values of the timestamp? Should you check when only one of `seconds` or `nanos` is `nil`?

File: lib/fluent/plugin/out_google_cloud.rb (hunk 2 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Changed.

File: lib/fluent/plugin/out_google_cloud.rb (hunk 3 above)
Comment (qingling128) [inline lib/fluent/plugin/out_google_cloud.rb]: Added more tests and changed the behavior a bit to be exactly aligned to how Logging API handles the cases (See details in the comments).

File: test/plugin/test_out_google_cloud.rb [hunk 4]
```diff
@@ -191,6 +191,30 @@ def test_parse_severity
...
+    ].each_with_index do |input, index|
+      setup_logging_stubs do
+        d = create_driver
+        @logs_sent = []
+        d.emit('message' => log_entry(0), 'timestamp' => input)
+        d.run
+      end
+      verify_log_entries(1, COMPUTE_PARAMS) do |entry|
+        assert_equal input, entry['metadata']['timestamp'],
+                     "Index #{index} failed with entry: #{entry.inspect}"
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud.rb]: [Optional] I wouldn't bother with the index here -- the entry is enough.

File: test/plugin/test_out_google_cloud.rb [hunk 5]
```diff
@@ -191,6 +191,30 @@ def test_parse_severity
     assert_equal('DEFAULT', test_obj.parse_severity('er'))
   end
 
+  def test_non_integer_timestamp
+    setup_gce_metadata_stubs
+    time = Time.now
+    [
+      { 'seconds' => nil, 'nanos' => nil },
+      { 'seconds' => nil, 'nanos' => time.tv_nsec },
+      { 'seconds' => nil, 'nanos' => time.tv_nsec },
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud.rb]: Any reason for the duplication?

File: test/plugin/test_out_google_cloud.rb (hunk 4 above)
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud.rb]: Ditto in `test_out_google_cloud_grpc.rb`...

Comment (qingling128) [inline test/plugin/test_out_google_cloud.rb]: But neither `entry` nor `expected` is sufficient to decide which test is failing for the grpc side.

Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud.rb]: Ah, you're right. Then perhaps use the stringified input instead of the index in both places?

Comment (qingling128) [inline test/plugin/test_out_google_cloud.rb]: Changed.

File: test/plugin/test_out_google_cloud_grpc.rb [hunk 6]
```diff
@@ -155,6 +155,32 @@ def test_struct_payload_non_utf8_log
...
+    }.each do |input, expected|
+      setup_logging_stubs do
+        d = create_driver
+        @logs_sent = []
+        d.emit('message' => log_entry(0), 'timestamp' => input)
+        d.run
+      end
+      verify_log_entries(1, COMPUTE_PARAMS) do |entry|
+        assert_equal expected, entry['metadata']['timestamp'],
+                     "Test with timestamp input '#{input}' failed with " \
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: s/timestamp input/timestamp/

Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud_grpc.rb]: Let's do the same in `test_out_google_cloud.rb`.

Comment (qingling128) [inline test/plugin/test_out_google_cloud_grpc.rb]: Changed.

Comment (qingling128) [inline test/plugin/test_out_google_cloud_grpc.rb]: Done.

File: test/plugin/test_out_google_cloud.rb [hunk 7]
```diff
@@ -208,7 +208,9 @@ def test_non_integer_timestamp
         d.run
       end
       verify_log_entries(1, COMPUTE_PARAMS) do |entry|
-        assert_equal timestamp, entry['metadata']['timestamp'], entry
+        assert_equal timestamp, entry['metadata']['timestamp'],
+                     "Test with timestamp '#{timestamp}' failed with " \
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud.rb]: Nit: s/failed with/failed for/. Also in the other file.

File: test/plugin/test_out_google_cloud.rb [hunk 8]
```diff
@@ -191,6 +191,30 @@ def test_parse_severity
...
+    ].each do |timestamp|
+      setup_logging_stubs do
+        d = create_driver
+        @logs_sent = []
+        d.emit('message' => log_entry(0), 'timestamp' => timestamp)
+        d.run
+      end
+      verify_log_entries(1, COMPUTE_PARAMS) do |entry|
+        assert_equal timestamp, entry['metadata']['timestamp'],
+                     "Test with timestamp '#{timestamp}' failed with " \
```
Comment (igorpeshansky) [inline test/plugin/test_out_google_cloud.rb]: Nit: s/failed with/failed for/. Also in the other file.

Comment (qingling128) [inline test/plugin/test_out_google_cloud.rb]: Changed.

**Comment by igorpeshansky** [comment]
LGTM :shipit:

**Pull request merged**